src/main/java/org/opensearch/security/transport/SecurityInterceptor.java

```diff
@@ -219,30 +224,46 @@ private void ensureCorrectHeaders(final Object remoteAdr, final User origUser, f
             getThreadContext().putHeader(ConfigConstants.OPENDISTRO_SECURITY_ORIGIN_HEADER, Origin.LOCAL.toString());
         }
 
+        TransportAddress transportAddress = null;
         if (remoteAdr != null && remoteAdr instanceof TransportAddress) {
-
             String remoteAddressHeader = getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS_HEADER);
-
             if(remoteAddressHeader == null) {
-                getThreadContext().putHeader(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS_HEADER, Base64Helper.serializeObject(((TransportAddress) remoteAdr).address()));
+                transportAddress = (TransportAddress) remoteAdr;
             }
         }
 
+        // we put headers as transient for same node requests
+        if (isSameNodeRequest) {
 
-        String userHeader = getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_USER_HEADER);
+            if(transportAddress != null) {
```

Comment:
For my knowledge, what's this `transportAddress`? Is this the same as `userHeader`?

Reply:
its the variable that holders the address of the remote cluster node (could be same cluster node too, but its store as remoteAddress in thread context)